### DISCUSSION TITLE
nvme performance

### DISCUSSION THREAD

**Discussion by rusticek** [moosefs/moosefs#604] [General]
After many years of running moosefs cluster with dosen of 7200rpm HDD's, primary for backups and archive, i build new cluster using NVMe and i think there is something wrong with the performance.

running version 3.0.117
1x master - EPYC 7302, 128GB RAM
2x chunkserver - EPYC 7713, 512GB RAM, 3x Samsung PM9A3
everything connected 2x10Gbit (LACP), MTU 1500

only changes to mfsmaster.cfg are
`CHUNKS_UNIQUE_MODE = 1`
`ATIME_MODE = 4`

Tests using FIO [from this site](https://cloud.google.com/compute/docs/disks/benchmarking-pd-performance-linux) are performed from mfsmaster machine, mounted like this
`mfsmaster: /mnt/mfs moosefs defaults,_netdev,nonempty,mfsdelayedinit,noatime,mfspassword=XXX,mfssubfolder=/stor-nvme1/ 0 0`

seq write - fine
`write: IOPS=1928, BW=1932MiB/s (2026MB/s)(566GiB/300275msec); 0 zone resets`
seq read - fine
`read: IOPS=3969, BW=3972MiB/s (4165MB/s)(1164GiB/300101msec)`
rand write - not fine
`write: IOPS=46.3k, BW=181MiB/s (190MB/s)(53.0GiB/300004msec); 0 zone resets`
rand read - not fine at all
`read: IOPS=42.7k, BW=167MiB/s (175MB/s)(48.9GiB/300005msec)`

To be clear. I do know that the MFS benefits from parallel operations. Iam not expecting milions IOPS, but iam surprised this is less then performance of one SATA SSD (random r/w).

**Comment by chogata** [comment]
What is your network latency (ping time)?

With random reads and writes you will always be constrained by network latency and the need to communicate with master server. This is a network filesystem, so a single operation is at first constrained by network speed first, and then by hdd speed, not in reverse. Multiple parallel operations speed the overall system up, because we use different machines at the same time. Single threaded random reads and writes will not benefit from NVMe drives - this will start to be visible as an improvement when you use many parallel threads.
Sequential writes and reads are faster, because a) they don't need to communicate with the master so much (the system doesn't have to ask about chunk locations so often) b) with read we have the read ahead mechanism too. Sequential parallel operations are the fastest, obviously :)

**Comment by rusticek** [comment]
rtt min/avg/max/mdev = 0.062/0.106/0.440/0.067 ms

I did some debuging and find out the mfsmaster process is at 80-90% CPU usage.
i moved mfsmaster to Ryzen 9 7950X3D and random writes and reads are at ~80k IOPS.

**Comment by chogata** [comment]
Well, the difference in CPU performance is understandable: 

https://www.cpubenchmark.net/cpu.php?cpu=AMD+EPYC+7302&id=3650 - single thread rating is 2019
vs
https://www.cpubenchmark.net/cpu.php?id=5234&cpu=AMD+Ryzen+9+7950X3D - single thread rating is 4149
so two times better. Master is a single threaded process, we always recommend low number of cores/high clock types of processors. Those two have the same number of cores, but single core performance difference is big.
With a faster CPU master queries will be faster too. Even network response time will be better with better CPU.

It's a bit strange that one random read/write test was causing 80-90% CPU usage on EPYC, it's not a bad CPU. Unless you were doing parallel tests? Or had a different profile for the CPU (energy saving and not performance)?

Also, I just noticed you still test 3.0.117. Test MooseFS 4 :)
